src/fpump.jl

```diff
@@ -0,0 +1,312 @@
+
+"""
+    generate_real_nlp(m,sol)
+
+Generate the orignal nlp and get the objective for that
+"""
+function generate_real_nlp(m, sol)
+    rmodel = Model(solver=m.nl_solver)
+    lb = [m.l_var; -1e6]
+    ub = [m.u_var; 1e6]
+
+    @variable(rmodel, lb[i] <= rx[i=1:m.num_var] <= ub[i])
+    for ni=1:m.num_int_bin_var
+        i = m.int2var_idx[ni]
+        JuMP.fix(rx[i], sol[i])
+    end
+
+    # define the objective function
+    obj_expr = MathProgBase.obj_expr(m.d)
+    expr_dereferencing!(obj_expr, rmodel)
+    JuMP.setNLobjective(rmodel, m.obj_sense, obj_expr)
+
+    # add all constraints
+    for i=1:m.num_constr
+        constr_expr = MathProgBase.constr_expr(m.d,i)
+        expr_dereferencing!(constr_expr, rmodel)
+        JuMP.addNLconstraint(rmodel, constr_expr)
+    end
+
+    status = solve(rmodel)
+    return status, getvalue(rx), getobjectivevalue(rmodel)
+end
+
+"""
+    add!(t::TabuList, m, sol)
+
+Add a solution to the tabu list (includes only the discrete variables)
+"""
+function add!(t::TabuList, m, sol)
+    t.sols[t.pointer] = [sol[m.int2var_idx[i]] for i=1:m.num_int_bin_var]
+    t.pointer += 1
+    if t.pointer > t.length
+        t.pointer = 1
+    end
+end
+
+"""
+    fpump(m)
+
+Run the feasibility pump 
+"""
+function fpump(m)
+    srand(1)
+
+    start_fpump = time()
+    nlp_sol = m.solution
+    nlp_obj = 1 # should be not 0 for while
+    c = 1
+    tabu_list = TabuList()
+    mip_sols = Dict{UInt64,Bool}()
+    tabu_list.length = 30
+    tabu_list.pointer = 1
+    tabu_list.sols = []
+    for i=1:tabu_list.length
+        push!(tabu_list.sols, NaN*ones(m.num_int_bin_var)) 
+    end
+
+    # construct the linear constraints as a vector of Aff once
+    if m.num_l_constr > 0
+        aff = construct_affine_vector(m)
+    end
+
+    fix = false
+    nlp_status = :Error
+    iscorrect = false
+    tl = m.options.feasibility_pump_time_limit
+    while !isapprox(nlp_obj,0.0, atol=atol) && time()-start_fpump < tl 
+        # generate a mip or just round if no linear constraints
+        if m.num_l_constr > 0
+            mip_status, mip_sol = generate_mip(m, nlp_sol, aff, tabu_list) 
+        else
+            # if no linear constraints just round the discrete variables
+            mip_sol = copy(nlp_sol)
+            mip_status = :Optimal
+            for vi=1:m.num_int_bin_var
+                vidx = m.int2var_idx[vi]
+                mip_sol[vidx] = round(mip_sol[vidx])
+            end
+        end
+        if mip_status != :Optimal
+            warn("MIP couldn't be solved to optimality")
+            break
+        end
+        
+        # If a cycle is detected which wasn't able to prevent by the tabu list (maybe too short)
+        if haskey(mip_sols, hash(mip_sol))
+            warn("Cycle detected")
+            break
+        end
+        add!(tabu_list, m, mip_sol)
+        mip_sols[hash(mip_sol)] = true
+
+        nlp_status, nlp_sol, nlp_obj = generate_nlp(m, mip_sol)
+        if nlp_status != :Optimal
```

Comment:
in the paper they suggested random restarts to mitigate this issue.

Reply:
Yeah might be reasonable to think about how to randomize it. Randomize only x or y or both?
Where x is discrete and y is continuous.

Reply:
It can be fairly generic route, before running solve on an NLP model.  For all variables pick a value uniformly at random between lb and ub and use `setvalue(x,v)` to assign the variable to this value.  NLP solvers are senative to the starting point, so this can help a lot of avoid converging to infeasible and suboptimal points.

Reply:
See  `num_resolve_at_*` options [here](https://www.coin-or.org/Bonmin/option_pages/options_list_bonmin.html#sec:Nonconvexproblems).